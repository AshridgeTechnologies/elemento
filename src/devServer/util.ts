import fs from 'fs'
import axios from 'axios'

export async function getModule(moduleUrl: string, moduleLocalPath: string, moduleName: string) {
    const moduleAlreadyDownloaded = fs.existsSync(moduleLocalPath)
    const moduleDate = moduleAlreadyDownloaded ? fs.statSync(moduleLocalPath).mtime : new Date(0)
    try {
        const resp = await axios.get(moduleUrl, {
            headers: {
                'If-Modified-Since': moduleDate.toUTCString(),
                'Cache-Control': 'max-age=31536000'
            },
            validateStatus: status => status >= 200 && status <= 304
        })
        if (resp.status === 304) {
            console.info(`Latest ${moduleName} already downloaded from`, moduleUrl)
            return true
        }
        if (resp.status !== 200) {
            throw new Error(resp.status + ' ' + resp.statusText)
        }
        const moduleContents = await resp.data
        await fs.promises.writeFile(moduleLocalPath, moduleContents)
        console.info(`Downloaded latest ${moduleName} from`, moduleUrl)
        return true
    } catch (e) {
        console.warn(`Cannot download latest ${moduleName} from`, moduleUrl, (e as Error).message)
        if (moduleAlreadyDownloaded) {
            console.warn(`Using previously downloaded version of ${moduleName}`)
            return true
        } else {
            return false
        }
    }
}