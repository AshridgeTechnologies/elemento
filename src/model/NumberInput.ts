import Element from './Element'
import {ElementId, ElementType, PropertyValueType} from './Types'
import BaseInputElement from './BaseInputElement'

export type Properties = {
    readonly initialValue?: PropertyValueType<number>,
    readonly label?: PropertyValueType<string>
}
export default class NumberInput extends BaseInputElement<Properties> implements Element {

    constructor(
        id: ElementId,
        name: string,
        properties: Properties) {
        super(id, name, properties)
    }

    static is(element: Element): element is NumberInput {
        return element.constructor.name === this.name
    }

    kind = 'NumberInput' as ElementType
}