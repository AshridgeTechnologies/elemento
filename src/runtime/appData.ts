import React from 'react'
import create from 'zustand'
import {stateProxy} from './stateProxy'
import DataStore, {Changes} from './DataStore'
import createContext from 'zustand/context'

type StoreType = {
    store: DataStore,
    update: (path: string, changes: Changes, replace: boolean) => void
}

const baseStore = (set: (updater: (state: StoreType) => object) => void): StoreType => ({
    store: new DataStore({app: {}}),
    update(path: string, changes: Changes, replace = false) {
        set((state: StoreType) => {
            const store = state.store
            const updatedStore = store.update(path, changes, replace)
            return {store: updatedStore}
        })
    }
})

const { Provider, useStore } = createContext<StoreType>()

const createStore = () => create(baseStore)

const useObjectStateWithDefaults = (path: string, initialValuesAndDefaults: object) => {
    const selectState = (state: any) => [state.store.select(path), state.update]
    const compareOnlyState = (a: any[], b: any[]) => a[0] === b[0]
    const [existingStateAtPath, updateFn] = useStore(selectState, compareOnlyState)
    return stateProxy(path, existingStateAtPath, initialValuesAndDefaults, updateFn)
}

const StoreProvider = ({children}: {children: React.ReactNode}) => React.createElement(Provider, {createStore, children})
export {StoreProvider, useObjectStateWithDefaults}
