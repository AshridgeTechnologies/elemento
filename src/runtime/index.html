<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="initial-scale=1, width=device-width" />
  <title>Elemento App</title>
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,500,700&display=swap"/>
</head>
<body>
  <div id="main"></div>
  <script src="https://unpkg.com/react@17/umd/react.development.js" crossorigin></script>
  <script src="https://unpkg.com/react-dom@17/umd/react-dom.development.js" crossorigin></script>
  <script src="runApp.ts" type="module"></script>
</body>
</html>