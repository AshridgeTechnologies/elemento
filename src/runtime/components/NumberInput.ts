import React, {ChangeEvent} from 'react'
import {TextField} from '@mui/material'
import {definedPropertiesOf} from '../../util/helpers'
import {valueOfProps} from '../runtimeFunctions'
import {proxyUpdateFnType} from '../stateProxy'
import {InputComponentState} from './InputComponentState'

type Properties = {state: {value?: number, _path: string, _controlValue: number | null, _update: proxyUpdateFnType}, label?: string}

export default function NumberInput({state, ...props}: Properties) {
    const {label} = valueOfProps(props)

    const optionalProps = definedPropertiesOf({label})
    const numericProps = {type: 'number', sx: { minWidth: 120, flex: 0 }}
    const {_path: path} = state
    const value = state._controlValue ?? ''
    const onChange = (event: ChangeEvent) => {
        const controlValue = (event.target as any).value
        const updateValue = controlValue !== '' ? Number(controlValue) : null
        state._update({value: updateValue})
    }

    return React.createElement(TextField, {
        id: path,
        variant: 'outlined',
        size: 'small',
        value,
        onChange,
        ...numericProps,
        ...optionalProps
    })
}

NumberInput.State = class State extends InputComponentState<number> {
    defaultValue = 0
}