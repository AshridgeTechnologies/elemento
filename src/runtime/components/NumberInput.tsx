import React, {ChangeEvent, FocusEvent} from 'react'
import {TextField} from '@mui/material'
import {definedPropertiesOf} from '../../util/helpers'
import {valueOfProps} from '../runtimeFunctions'
import InputComponentState from './InputComponentState'
import {NumberType} from '../types'
import {isNil, pick} from 'ramda'
import BigNumber from 'bignumber.js'
import DecimalType from '../types/DecimalType'
import BaseType from '../types/BaseType'
import {
    BaseInputComponentProperties,
    getLabelWithRequired,
    inputElementProps,
    propsForInputComponent,
    sxPropsForFormControl
} from './ComponentHelpers'
import {useObject} from '../appStateHooks'

type Properties = BaseInputComponentProperties

const decPlaces = (dataType: BaseType<any, any> | undefined) => dataType?.kind === 'Decimal' ? (dataType as DecimalType).decimalPlaces : undefined

const dataTypeProps = (dataType: BaseType<any, any> | undefined) => {
    const props = definedPropertiesOf(pick(['max', 'min'], dataType ?? {}))
    props.step = decPlaces(dataType) === 0 ? 1 : undefined
    return props
}

export default function NumberInput({path, ...props}: Properties) {
    const {label, readOnly, show, styles = {}} = valueOfProps(props)
    const sx = sxPropsForFormControl(styles, show, {minWidth: 120, flex: 0})

    const state: NumberInputState = useObject(path)
    const {dataValue, dataType} = state

    const formatValue = () => {
        if (isNil(dataValue)) return ''
        const dp = decPlaces(dataType)
        return dp !== undefined ? dataValue.toFixed(dp) : dataValue.toString()
    }
    const value = state.controlValue ?? formatValue()
    const labelWithRequired = getLabelWithRequired(dataType, label)
    const optionalProps = definedPropertiesOf({label: labelWithRequired})
    const inputComponentProps = propsForInputComponent(dataType, styles)
    const inputProps = inputElementProps(styles, readOnly, dataTypeProps(dataType))

    const error = state.errorsShown && !state.valid
    const helperText = state.errorsShown && state.errors ? (state.errors as string[]).join('.  ') : undefined

    const onChange = (event: ChangeEvent) => {
        const controlValue = (event.target as any).value
        const updateValue = () => {
            if (controlValue === '') return null
            if (dataType?.kind === 'Decimal') return new BigNumber(controlValue)
            return Number(controlValue)
        }
        state._setValues(updateValue(), controlValue)
    }
    const onBlur = (_event: FocusEvent) => state._setBlurred()

    return React.createElement(TextField, {
        id: path,
        variant: 'outlined',
        size: 'small',
        type: 'number',
        value,
        error,
        helperText,
        onChange,
        onBlur,
        sx,
        ...inputProps,
        ...inputComponentProps,
        ...optionalProps
    })
}

export class NumberInputState extends InputComponentState<number | BigNumber, NumberType | DecimalType>  {
    defaultValue = 0

    _setValues(value: number | BigNumber | null, editedValue: string) {
        this.updateState({value, editedValue})
    }
}

NumberInput.State = NumberInputState
