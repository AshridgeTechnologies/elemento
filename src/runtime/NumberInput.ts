import React, {ChangeEvent} from 'react'
import {TextField} from '@mui/material'
import {definedPropertiesOf} from '../util/helpers'
import {updateState, useObjectState} from './appData'

type Properties = {path: string, label?: string, initialValue?: number}

export default function NumberInput({path, initialValue = 0, label}: Properties) {
    const state = useObjectState(path)
    const optionalProps = definedPropertiesOf({label})
    const value = (state?.value !== undefined) ? state.value : initialValue
    const onChange = (event: ChangeEvent) => updateState(path, {value: Number((event.target as any).value) })

    return React.createElement(TextField, {
        id: path,
        type: 'text',
        variant: 'outlined',
        size: 'small',
        value,
        onChange,
        ...optionalProps
    })
}
