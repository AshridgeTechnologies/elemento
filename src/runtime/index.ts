import * as components from './components'
import {globalFunctions} from './globalFunctions'
export {components, globalFunctions}

export {default as React} from 'react'
export {default as ReactDOM} from 'react-dom/client'
export {default as appFunctions} from './appFunctions'
export {useGetObjectState, useObjectState} from './appData'
export {asArray, codeGenerationError, parentPath} from './runtimeFunctions'
export {default as AppRunner} from './AppRunner'
export {run} from './run'
