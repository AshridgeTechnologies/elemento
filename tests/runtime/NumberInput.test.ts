/**
 * @jest-environment jsdom
 */

import {createElement} from 'react'
import NumberInput from '../../src/runtime/NumberInput'
import {snapshot, testContainer} from '../util/testHelpers'
import userEvent from '@testing-library/user-event'
import {proxify, useStore} from '../../src/runtime/appData'

test('NumberInput element produces output with properties supplied',
    snapshot(createElement(NumberInput, {state: proxify({value: 27}, 'app.page1.width'), label: 'Width'}))
)

test('NumberInput element produces output with default values where properties omitted',
    snapshot(createElement(NumberInput, {state: proxify({value: 0}, 'app.page1.height')}))
)

test('NumberInput shows value from the state supplied', () => {
    let container = testContainer(createElement(NumberInput, {state: proxify({value: 27}, 'app.page1.widget1')}))
    expect(container.querySelector('input[id="app.page1.widget1"]').value).toBe('27')
})

test('NumberInput element produces output with properties supplied as state objects', () => {
    let container = testContainer(createElement(NumberInput, {
        state: proxify({value: 27}, 'app.page1.widget1'),
        label: proxify({value: 'Item Number'}, 'path.x')
    }))
    expect(container.querySelector('label[for="app.page1.widget1"]').innerHTML).toBe('Item Number')
})

test('NumberInput shows empty value when state value is absent', () => {
    let container = testContainer(createElement(NumberInput, {state: proxify({}, 'app.page1.widget1')}))
    expect(container.querySelector('input[id="app.page1.widget1"]').value).toBe('')
})

test('NumberInput shows empty value when state value is set to undefined', () => {
    let container = testContainer(createElement(NumberInput, {state: proxify({value: undefined}, 'app.page1.widget1')}))
    expect(container.querySelector('input[id="app.page1.widget1"]').value).toBe('')
})

test('NumberInput stores updated values in the app store section for its path', async () => {
    let container = testContainer(createElement(NumberInput, {state: proxify({value: 27}, 'app.page1.sprocket')}))
    const inputEl = container.querySelector('input[id="app.page1.sprocket"]')
    const user = userEvent.setup()
    await user.type(inputEl, '6')
    expect((useStore.getState() as any).app.page1.sprocket).toStrictEqual({value: 276})
    //expect(inputEl.value).toBe('66')
} )

test('NumberInput stores undefined value in the app store when cleared', async () => {
    let container = testContainer(createElement(NumberInput, {state: proxify({value: 27}, 'app.page1.sprocket')}))
    const inputEl = container.querySelector('input[id="app.page1.sprocket"]')
    const user = userEvent.setup()
    await user.clear(inputEl)
    expect((useStore.getState() as any).app.page1.sprocket).toStrictEqual({value: undefined})
    //expect(inputEl.value).toBe('66')
} )
