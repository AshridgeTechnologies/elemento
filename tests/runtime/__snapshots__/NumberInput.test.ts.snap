// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NumberInput element produces output with default values where properties omitted 1`] = `
<div
  className="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
>
  <div
    className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
    onClick={[Function]}
  >
    <input
      aria-invalid={false}
      autoFocus={false}
      className="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
      disabled={false}
      id="app.page1.height"
      onAnimationStart={[Function]}
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      required={false}
      type="text"
      value={0}
    />
    <fieldset
      aria-hidden={true}
      className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
    >
      <legend
        className="css-1a1ihhj"
      >
        <span
          className="notranslate"
          dangerouslySetInnerHTML={
            Object {
              "__html": "&#8203;",
            }
          }
        />
      </legend>
    </fieldset>
  </div>
</div>
`;

exports[`NumberInput element produces output with properties supplied 1`] = `
<div
  className="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
>
  <label
    className="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary MuiFormLabel-filled css-1kty9di-MuiFormLabel-root-MuiInputLabel-root"
    data-shrink={true}
    htmlFor="app.page1.width"
    id="app.page1.width-label"
  >
    Width
  </label>
  <div
    className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
    onClick={[Function]}
  >
    <input
      aria-invalid={false}
      autoFocus={false}
      className="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
      disabled={false}
      id="app.page1.width"
      onAnimationStart={[Function]}
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      required={false}
      type="text"
      value={27}
    />
    <fieldset
      aria-hidden={true}
      className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
    >
      <legend
        className="css-186xcr5"
      >
        <span>
          Width
        </span>
      </legend>
    </fieldset>
  </div>
</div>
`;
